src/main.rs

```diff
@@ -595,105 +601,137 @@ fn verify_files_main(
-            vc_is_valid.run_solver(options, &limits_ref, &ctx, &mut translate, &slice_vars)?;
+            // 11. Translate to Z3
+            let ctx = mk_z3_ctx(options);
+            let smt_ctx = SmtCtx::new(&ctx, &tcx);
+            let mut translate = TranslateExprs::new(&smt_ctx);
+            let mut vc_is_valid = vc_is_valid.into_smt_vc(&mut translate);
 
-        server
-            .handle_vc_check_result(name, verify_unit.span, &mut result, &mut translate)
-            .map_err(VerifyError::ServerError)?;
+            // 12. Simplify
+            if !options.no_simplify {
+                vc_is_valid.simplify();
+            }
 
-        // If requested, write the SMT-LIB output.
-        result.write_smtlib(options, name)?;
+            // 13. Create Z3 solver with axioms, solve
+            let mut result =
+                vc_is_valid.run_solver(options, &limits_ref, &ctx, &mut translate, &slice_vars)?;
 
-        if options.z3_trace {
-            info!("Z3 tracing output will be written to `z3.log`.");
-        }
+            server
+                .handle_vc_check_result(name, verify_unit.span, &mut result, &mut translate)
+                .map_err(VerifyError::ServerError)?;
+
+            // If requested, write the SMT-LIB output.
+            result.write_smtlib(options, name)?;
 
-        // Handle reasons to stop the verifier.
-        match result.prove_result {
-            ProveResult::Unknown(ReasonUnknown::Interrupted) => {
-                return Err(VerifyError::Interrupted)
+            if options.z3_trace {
+                info!("Z3 tracing output will be written to `z3.log`.");
             }
-            ProveResult::Unknown(ReasonUnknown::Timeout) => return Err(LimitError::Timeout.into()),
-            _ => {}
-        }
 
-        // Increment counters
-        match result.prove_result {
-            ProveResult::Proof => num_proven += 1,
-            ProveResult::Counterexample(_) | ProveResult::Unknown(_) => num_failures += 1,
-        }
+            // Handle reasons to stop the verifier.
+            match result.prove_result {
+                ProveResult::Unknown(ReasonUnknown::Interrupted) => {
+                    return Err(VerifyError::Interrupted);
+                }
+                ProveResult::Unknown(ReasonUnknown::Timeout) => {
+                    return Err(LimitError::Timeout.into())
+                }
+                _ => {}
+            }
+
+            // Increment counters
+            match result.prove_result {
+                ProveResult::Proof => num_proven += 1,
+                ProveResult::Counterexample(_) | ProveResult::Unknown(_) => num_failures += 1,
+            }
+
+            Ok(())
+        })();
+        match res {
+            Ok(_) => Ok(()),
+            Err(VerifyError::LimitError(_)) | Err(VerifyError::Interrupted) => {
+                wrapping_up = true;
+                server
+                    .handle_not_checked(verify_unit.span)
```

Comment:
Are you discarding the reason for stopping here? Was it already handled previously? Can it happen that both `handle_vc_check_result` and `handle_not_checked` are called for a proc? What happens then?

Reply:
They can not be called together for the same proc because all limit checks happen before the `handle_vc_check_result` call. The error shouldn't be discarded, this problem is solved in the following commit.